### PR TITLE
KAFKA-5192 WindowStore range scan KIP-155

### PR DESCRIPTION
**Pull request by xvrl**
Implements range scan for keys in windowed and session stores

Modifies caching session and windowed stores to use segmented cache keys.
Cache keys are internally prefixed with their segment id to ensure key ordering in the cache matches the ordering in the underlying store for keys spread across multiple segments.
This should also result in fewer cache keys getting scanned for queries spanning only some segments.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/ReadOnlySessionStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/ReadOnlySessionStore.java:44]: I guess we need javadoc for these new methods on the public interface

File: streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractMergedSortedCacheStoreIterator.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractMergedSortedCacheStoreIterator.java]: nit: `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java]: These should use `serdes.rawKey(...)` as has been done in: https://github.com/apache/kafka/pull/2963

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStore.java]: This method is basically identical to `fetch(final K key)`.  Maybe create an interface, `Fetcher` (or similar) with 2 impls, one that delegates to `store.fetch(from, to)` and the other that delegates to `store.fetch(key)`. Then make both `fetch` methods delegate to another internal fetch method passing in the appropriate impl?

Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStore.java]: nit: `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java]: nit: `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java]: Same as in `CachingSessionStore`. Both `fetch` methods are basically identical.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheKeyValueStoreIterator.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheKeyValueStoreIterator.java]: nit: `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java]: can be package private

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java]: nit: `final` here and elsewhere

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: Could we make the other `fetch` method delegate to this one?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java]: Could we remove this method and just have all the `fetch` calls use the other one?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java]: As noted above - can we remove the other overload of this method and just use this one?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java]: Same as above

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStore.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStore.java]: sounds reasonable

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java]: my only concern with removing the single key method would be the potential performance impact of having to check both `<=` and `>=` for each value as opposed to just `=` for single key lookups.

**Comment by xvrl** [comment]
@dguy addressed comments in your initial review, and also expanded test coverage.
After some testing I realized we needed to have the cache also respect the segmentation of the underlying store to have a consistent ordering between cache and underlying store, when querying for key ranges across segments.

This led me to change the caching stores to prefix every cache key with a segment id internally. One benefit of that, is that time range queries will now only scan relevant segments in the cache as well, potentially reducing the number of cache key to scan.

There were also some gotchas with respect to key ranges which were a bit tricker, so I added the upper/lowerRangeFixedSize to keep the optimized ranges for single key queries

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: underlying byte ranges differ significantly for single keys vs key range, so I kept both methods

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java]: Ok - i guess i'd like to understand what the performance impact is.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java]: nit: we generally make all params `final`

I guess there are other places, too

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java]: uggh - i think i saw you mentioned something about the checkstyle violation todo with generic types. Maybe we should change that rule? All caps is pretty ugly and Hard to read.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/FilteredCacheIterator.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/FilteredCacheIterator.java:34]: Could we not pass this iterator in via the ctor? Wouldn't need the `CacheFunction` anymore and would mean that we could write tests just for this if need be

File: streams/src/main/java/org/apache/kafka/streams/state/internals/KeyValueIterators.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/KeyValueIterators.java]: nit: can be package-private

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreIterator.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreIterator.java:34]: All the subclasses of `AbstractMergedSortedCacheStoreIterator` need `serdes` any reason why you've pulled the field into the subclasses? Something to do with generics?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java:26]: Can you add a unit test for this class?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java]: can we add a unit test for this class?

Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java]: does it need to be public?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java]: nit: package-private?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java]: nit: package-private?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java]: is this and `lowerRange` identical to those in `SessionKeySchema`? If so then maybe we should extract them to avoid code duplication?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java]: nit: `private` ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java:28]: would be good to add some unit tests for this

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java]: can we extract these out into separate tests? The second and 3rd "tests" are testing a different function. I'd prefer this was three tests.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: Create another test for this?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: Same here. Would prefer to see at least 2 tests, one for each of the `fetch` methods

File: streams/src/test/java/org/apache/kafka/streams/state/internals/SessionKeySchemaTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/SessionKeySchemaTest.java]: would prefer to see separate tests for each of these. Describing what they are testing

File: streams/src/test/java/org/apache/kafka/streams/state/internals/WindowKeySchemaTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/WindowKeySchemaTest.java]: as above

**Review by dguy — COMMENTED** [review]
Thanks @xvrl left some more comments

File: streams/src/main/java/org/apache/kafka/streams/state/internals/FilteredCacheIterator.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/FilteredCacheIterator.java:34]: I'd rather not, since the caller otherwise would have to know that both of those iterators need to share the same underlying iterator. I'd rather have this class encapsulate that.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreIterator.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreIterator.java:34]: AbstractMergedSortedCacheStoreIterator doesn't need the serdes, since it delegates all the serialization / deserialization through its abstract methods. The main motivation was this bug https://github.com/apache/kafka/pull/3016 which was caused by calling serdes directly within the abstract class.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java]: maxSuffix is different

File: streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java]: I can make it protected, it's being used in the overridden methods above, I think IntelliJ wasn't smart enough to figure that out.

Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java]: yeah - probably

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java]: I'm happy to use a single method for now. In the essence of time. If there is a clear need for both, I'll do a follow-up PR.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java]: removed

File: streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java]: could still probably extract a method that removes almost all of the duplication?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java:28]: this is not actually a new class, it used to be `WrappedWindowStoreIterator`. I just renamed it since it's technically not an Iterator anymore, but instead provides wrapped iterators.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/SessionKeySchemaTest.java
Comment (xvrl) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/SessionKeySchemaTest.java]: will do

File: streams/src/main/java/org/apache/kafka/streams/state/ReadOnlySessionStore.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/ReadOnlySessionStore.java:44]: done

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java:26]: done

**Comment by xvrl** [comment]
@dguy I believe I addressed all your comments

**Review by dguy — APPROVED** [review]
Thanks @xvrl - LGTM. @guozhangwang, @mjsax, @enothereska, @bbejeck can we please get a second review?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java:52]: We may need to quantify how much perf impact it may incur for such allocation calls (note we need to trigger once per cache put): this is not a comment for this PR specifically.

**Review by guozhangwang — APPROVED** [review]
LGTM! Big thanks to @xvrl for this huge contribution!

**Comment by guozhangwang** [comment]
Merged to trunk.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java
Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java:52]: agree, when I was implementing this, I was thinking we should switch our serde to ByteBuffers, so we could more easily work with array slices, since we often just pass around portions of the underlying array when working with composite keys such as windowed / segmented keys.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java:52]: Agreed, could you file a separate JIRA for this?

Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java:52]: done https://issues.apache.org/jira/browse/KAFKA-5966

**Pull request closed without merging**
